### PR TITLE
docs: Senna.js integration in DXP 

### PR DESCRIPTION
**Pull request by julien**
@matuzalemsteles and I worked on this to document Senna.js integration in DXP.
It probably still need some changes/improvements, but we think this is good for a first review

### CODE REVIEW

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:11]: > has the correct value

Should we mention what that value is? Off the top of my head, I don't know.

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:17]: You're missing a full-stop at the end of this sentence.

What makes us think that this might change in the future? (I know we're trying to get rid of globals, but this one seems particularly unlikely to go away soon.)

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:19]: "modules" → "module"

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:21]: Backticks around `SPATopHeadJSPDynamicInclude`.

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:44]: Really? So if I set it to `-0.9` it will still cache?

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:50]: "into memory" → "in memory"

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:54]: "it's" → "its"

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:60]: Need a full-stop at the end of this sentence.

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:66]: This sentence isn't all that clear; I think I'd just remove it.

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:68]: Can we call them "concepts" instead of "components"? (Otherwise they sound a bit like "UI components")

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:74]: At this is a link to a line number that may change, can we make it permanent by including a hash? (ie. hit "y" in the GitHub UI.)

Also, we have a run on sentence here, so put a full-stop after "within the Surface".

"Browser" → "browser" because it is not a proper noun or even a special concept in this document.

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:92]: These should all have backticks.

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:98]: No "the" before `RequestScreen`.

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:104]: Instead of:

> when everything is ready, style, scripts, and elements.

Do:

> when everything is ready (styles, scripts, and elements).

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:110]: Another run-on sentence.

> Resource tracking of `style` and `script` tags happens during navigation. `Screen` looks for those elements that are marked with the `data-senna-track` attribute within the content that is in a virtual document or in the DOM.
>
> There are two types of tracking for these attributes:

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:117]: Full-stop after "evaluated".

**Review by wincent — COMMENTED** [review]
Some little nits inline, but for me (as a non-Senna-expert) this is a very useful guide.

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:11]: Yes that's actually a good idea, the value should be `0` - or at least that's what I understand when I read [this](https://github.com/liferay/liferay-portal/blob/b619978445e3be09526803b27447b5ddab95ca6f/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/liferay/screen/ActionURLScreen.es.js#L53-L55)

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:17]: You read my mind, I actually thought that we might be able to get rid of this global - without even searching for usages :see_no_evil: 
As you said; this one will probably be hard to get rid of, so I might as well remove that.

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:44]: It [looks](https://github.com/liferay/liferay-portal/blob/021b09a2df5e71140dd6061497000db108f09048/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/liferay/app/App.es.js#L136-L138) like it will

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:54]: I always get that one wrong - it's a shame

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:68]: Yes of course, I just didn't come up with a better name - but it's true that "components" is confusing in this case.

File: dxp/sennajs_integration.md
Comment (wincent) [reply dxp/sennajs_integration.md:44]: Although [here](https://github.com/liferay/liferay-portal/blob/021b09a2df5e71140dd6061497000db108f09048/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/liferay/app/App.es.js#L171-L177), it looks like if you set a negative time `screenCachedExpired` will always be `true` (because `lastModifiedInterval` should always be non-negative), which is the same as having no cache.

So technically, unless I misunderstand how this works, the actual truth is that Senna will cache if the value is greater than `0`, not `-1`.

🤷

Comment (julien) [reply dxp/sennajs_integration.md:44]: Oh ... right.

Comment (julien) [reply dxp/sennajs_integration.md:44]: I'll fix that

**Comment by matuzalemsteles** [comment]
I'm not sure if I managed to give a clear understanding of how Senna.js works, but let me know if we can add more things, the idea is to pass on an understanding and not go into the methods so much. Does that make sense?

**Comment by wincent** [comment]
> I'm not sure if I managed to give a clear understanding of how Senna.js works, but let me know if we can add more things, the idea is to pass on an understanding and not go into the methods so much. Does that make sense?

I think it makes sense. The idea is not to explain every detail of the source code, but to give a high-level idea so that when somebody comes to look at the source, they have a "conceptual map" to find their way around. Without the "map", you have to reverse-engineer the thing in order to begin understanding it, and that's hard.

**Comment by matuzalemsteles** [comment]
> I think it makes sense. The idea is not to explain every detail of the source code, but to give a high-level idea so that when somebody comes to look at the source, they have a "conceptual map" to find their way around. Without the "map", you have to reverse-engineer the thing in order to begin understanding it, and that's hard.

I completely agree 🙂.

File: dxp/sennajs_integration.md
Comment (jbalsas) [inline dxp/sennajs_integration.md:115]: How are elements without the `data-senna-track` attribute treated by default? ;)

File: dxp/sennajs_integration.md
Comment (jbalsas) [inline dxp/sennajs_integration.md:72]: I don't see this explained anywhere... how many surfaces do we have in DXP? which one(s)? This is interesting because people might thing this is working in a way when it's working in a different way.

Comment (jbalsas) [reply dxp/sennajs_integration.md:72]: Once you get that, I think it's also interesting to add some diagram or explanation of what type of content are we getting back from DXP backend and what do we do with it.

Spoiler alert, we get full documents as if it were full refreshes (not partials), and then there's some things that get filtered or ignored... I think this would be interesting to write down.

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:115]: I think it's "permanent" but I'll look into the code and document it

Comment (matuzalemsteles) [reply dxp/sennajs_integration.md:115]: Yeah, if the element is in the `head` without the attribute it will be permanent Senna.js will not touch this element. But if the element is in `body` or specifically within the content of the page without the attribute, the behavior will be like a temporary.

File: dxp/sennajs_integration.md
Comment (matuzalemsteles) [reply dxp/sennajs_integration.md:72]: Yeah, I think that in the concept of Senna.js there was to be more than one Surface but for DXP only one is used during startup, I don't know exactly why. But yes I can document these details more on the DXP side, in fact, it will happen independently of DXP as well.

Comment (jbalsas) [reply dxp/sennajs_integration.md:72]: > Yeah, I think that in the concept of Senna.js there was to be more than one Surface but for DXP only one is used during startup, I don't know exactly why.

I only vaguely remember why now... maybe it can be found in DXP's git history, I'll try to see if I can find the specific commits tomorrow and if not, I'll simply share what I remember.

Comment (matuzalemsteles) [reply dxp/sennajs_integration.md:72]: > I only vaguely remember why now... maybe it can be found in DXP's git history, I'll try to see if I can find the specific commits tomorrow and if not, I'll simply share what I remember.

Thanks, I'll try to look for something in the commits too.

> Spoiler alert, we get full documents as if it were full refreshes (not partials), and then there's some things that get filtered or ignored... I think this would be interesting to write down.

Yeah, this is an important thing, I'll make that clearer.

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:115]: Thanks @matuzalemsteles I'll add that right now

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:72]: The only surface I found was `LiferaySurface` but I'll also search in the git history

Comment (jbalsas) [reply dxp/sennajs_integration.md:72]: `LiferaySurface` is a class. When the code run, we instantiate and register a number of surfaces which is what's in the diagram above. Don't go looking in the history for code, figure out which surfaces we create and manage in runtime and then check if we did anything different at any other point in time.

Comment (julien) [reply dxp/sennajs_integration.md:72]: I probably should have said that `LiferaySurface` is a class, but the only instance of that class being created is [here](https://github.com/liferay/liferay-portal/blob/3ec7c6e5d08fa594303f140c55407f7158ecccb3/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/liferay/app/App.es.js#L79). 
At one point we changed [this](https://github.com/liferay/liferay-portal/commit/70b1010a4a521b5a4f955ae281c9e5019faa643a) to prevent overwritting the `document.body.id` if it has one.

**Comment by matuzalemsteles** [comment]
I added some more information, let me know if I forgot any details.

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:74]: I'd be interested in knowing some more about what each of these does.

For example, without looking at the code, I don't know the difference between `beforeNavigate`, `startNavigate` and `beforeScreenFlip` (although I can make a guess, and make assumptions about the order they fire in).

Likewise, I don't know the order of or difference between `screenDeactivate` and `screenDispose` — again, I can take a guess, or look at the code, but it would be cool if it were spelled out for me.

Or another way to think of it, we might want to define what the basic terms mean:

- What "flipping" a screen means.
- What "activating" and "deactivating" (and "disposing") a screen means.

Of the terms, the only ones that have relatively clear meanings are "navigate" and "load", but even then we need to be clear about what they mean in the context of Senna.

**Review by wincent — COMMENTED** [review]
I don't want to let "perfect be the enemy of the good" here, so I think we should ship this soon. There is just one area I've commented on where I think there is a bit of a gap; I leave it up to you @julien and @matuzalemsteles to decide whether it is worth filling that gap.

File: dxp/sennajs_integration.md
Comment (jbalsas) [inline dxp/sennajs_integration.md:40]: Might be good to explain how this is done technically... via the OSGi header or by other means.

File: dxp/sennajs_integration.md
Comment (jbalsas) [inline dxp/sennajs_integration.md:54]: I think there's a typo in `spa.exluced.paths`. Hope it's `excluded` and we didn't miss it the whole time :D

File: dxp/sennajs_integration.md
Comment (jbalsas) [reply dxp/sennajs_integration.md:74]: Agreed... a high level description of the lifecycle events would be helpful. We have some information in the source code itself or in sennajs.com, so just putting together some of it here should be enough, I think.

**Review by jbalsas — APPROVED** [review]
Other than the request to expand a little bit on the meaning of the lifecycle methods, this looks good to me! 😍

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:72]: :+1: good catch

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md:93]: Is it true that `beforeScreenFlip` comes before `screenLoad`?

File: dxp/sennajs_integration.md
Comment (wincent) [inline dxp/sennajs_integration.md]: I think it was right before:

Screenshot 2020-10-05 -170739-PAc7RTj5@2x

This commit makes it look like this:

Screenshot 2020-10-05 -170704-7sFIeWNz@2x

(14 commits and counting; we had better squash all this before merging.)

Comment (julien) [reply dxp/sennajs_integration.md]: I'm going to fix this and squash, I don't know if it's me but each time I try to view the change on GitHub it's messed up

File: dxp/sennajs_integration.md
Comment (julien) [reply dxp/sennajs_integration.md:90]: no, it's not. I fixed that in https://github.com/liferay/liferay-frontend-guidelines/pull/223/commits/2e2b34b5b3726a1182a4a1b16801bf2977b29771

**Review by wincent — APPROVED** [review]
Seems everybody is happy with this, and the rendered output looks good, so let's get this in.

**Pull request merged**
